apps/mark-scan/backend/src/custom-paper-handler/state_machine.ts

```diff
@@ -1,267 +1,394 @@
+  StateNodeConfig,
+  interpret,
+  Interpreter,
+  Assigner,
+  PropertyAssigner,
+  ServiceMap,
+  StateSchema,
+} from 'xstate';
+import { switchMap, throwError, timeout, timer } from 'rxjs';
+import { Optional, assert } from '@votingworks/basics';
+import { SheetOf } from '@votingworks/types';
+import { singlePrecinctSelectionFor } from '@votingworks/utils';
+import { Workspace } from '../util/workspace';
+import { SimpleServerStatus } from './types';
+import {
+  PAPER_HANDLER_STATUS_POLLING_INTERVAL_MS,
+  PAPER_HANDLER_STATUS_POLLING_TIMEOUT_MS,
+} from './constants';
+import {
+  interpretScannedBallots,
+  isPaperInScanner,
+  isPaperReadyToLoad,
+  isPaperInOutput,
+  scanAndSave,
+  setDefaults,
+  printBallot as driverPrintBallot,
+} from './application_driver';
 
-function isPaperInInput(paperHandlerStatus: PaperHandlerStatus): boolean {
-  return (
-    paperHandlerStatus.paperInputLeftInnerSensor ||
-    paperHandlerStatus.paperInputLeftOuterSensor ||
-    paperHandlerStatus.paperInputRightInnerSensor ||
-    paperHandlerStatus.paperInputRightOuterSensor
-  );
+interface Context {
+  workspace: Workspace;
+  driver: PaperHandlerDriver;
+  pollingIntervalMs: number;
+  scannedImagePaths?: SheetOf<string>;
 }
 
-function isPaperReadyToLoad(paperHandlerStatus: PaperHandlerStatus): boolean {
-  return (
-    paperHandlerStatus.paperInputLeftInnerSensor &&
-    paperHandlerStatus.paperInputLeftOuterSensor &&
-    paperHandlerStatus.paperInputRightInnerSensor &&
-    paperHandlerStatus.paperInputRightOuterSensor
-  );
+function assign(arg: Assigner<Context, any> | PropertyAssigner<Context, any>) {
+  return xassign<Context, any>(arg);
 }
 
-function isPaperInScanner(paperHandlerStatus: PaperHandlerStatus): boolean {
-  return (
-    paperHandlerStatus.paperPreCisSensor ||
-    paperHandlerStatus.paperPostCisSensor ||
-    paperHandlerStatus.preHeadSensor ||
-    paperHandlerStatus.paperOutSensor ||
-    paperHandlerStatus.parkSensor ||
-    paperHandlerStatus.paperJam ||
-    paperHandlerStatus.scanInProgress
-  );
-}
+type PaperHandlerStatusEvent =
+  | { type: 'NO_PAPER_IN_FRONT' }
+  | { type: 'PAPER_READY_TO_LOAD' }
+  | { type: 'PAPER_INSIDE' }
+  | { type: 'VOTER_INITIATED_PRINT' }
+  | { type: 'BALLOT_PRINTED' }
+  | { type: 'SCANNING' }
+  | { type: 'PAPER_REMOVED' }
+  | { type: 'END' };
 
-function isPaperAnywhere(paperHandlerStatus: PaperHandlerStatus): boolean {
-  return (
-    isPaperInInput(paperHandlerStatus) || isPaperInScanner(paperHandlerStatus)
-  );
-}
+const debug = makeDebug('mark-scan:state-machine');
 
 export class PaperHandlerStateMachine {
-  private status: SimpleStatus = 'no_paper';
-  private lastActionInitiatedTime = 0;
-
-  constructor(private readonly driver: PaperHandlerDriver) {}
+  constructor(
+    private readonly driver: PaperHandlerDriver,
+    private readonly machineService: Interpreter<
+      Context,
+      any,
+      PaperHandlerStatusEvent,
+      any,
+      any
+    >
+  ) {}
```

Comment:
I know this class was here before you started, but I wonder if it would be better to match the pattern of apps/scan (constructing a plain object with methods) in order to have less cognitive overhead when looking at the two files.

Reply:
I think that's a valuable change to make. Would you be opposed if we keep this thin wrapper class in this PR and refactor after we hit the demo milestone? I don't think the cost of moving to the`apps/scan` pattern will increase much if we wait, and I think we should prioritize exception handling first.

Reply:
Yeah, fine to punt til later. I didn't realize this PR was just aiming for a demo. That helps explain why the state machine is so much simpler than `apps/scan`!